keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,157 @@
+
+### User Stories
+
+#### Story 1
+
+As a user of Kubernetes, I can query the services I can declaratively request using a Kubernetes native API. For example, using the command `kubectl get crds` where I can see services alongside other resources that can be created.
+
+#### Story 2
+
+As a user of Kubernetes, I can declaratively request an instance of a service using a custom resource. When the service is provided the means to use that service (e.g., credentials in a secret) are provided in a common and consistent manner. The same resource and secret can be used in clusters running in different locations and the way the service is provided may be different.
+
+#### Story 3
+
+As a cluster operator or application operator, I can discover controllers implementing the CRDs and secrets to support the application portability in my cluster.
+
+#### Story 4
+
+As a cluster operator or application operator, I can set default values and provider custom settings for a service. 
+
+### Implementation Details/Notes/Constraints
+
+To solve the two user stories there are two types of Kubernetes resources that can be leveraged.
+
+1. Custom resource definitions (CRDs) can be used to describe a service. The CRDs can be implemented by controllers for different environments and the list of installed CRDs can be queried to see what is supported in a cluster
+2. Secrets with a specific type and schema can be used to handle credentials and other relevant information for services that have them (e.g., a database). Not all services will require a secret (e.g., DNS)
+
+This subproject will list and document the resources and how controllers can implement them. This provides for interoperability including that for controllers and other tools, like validators, and a canonical listing held by a vendor neutral party.
+
+In addition to the resources, this subproject will also provide a controller implementing the defined services to support testing, providing an example implementation, and to support other Kubernetes subprojects (e.g., Minikube). Controller produced by this project are _not_ meant to be used in production.
+
+3rd party controllers implementing the CRDs and secrets can use a variety of methods to implement the service handling. This is where the Kubernetes Service Catalog can be an option. This subproject will not host or support 3rd party controllers but will list them to aide in users discovering them. This is in support of the 3rd user story.
+
+To support custom settings for services by a service provider and to add the ability to add default settings (user story 4) we are considering a pattern of using CRDs for a controller with configuration on a cluster wide and namespace based level. An example of this in existence today is the cert manager issuer and cluster issuer resources. How to support this pattern will be worked out as part of the process in the next step of building a working system. This pattern is tentative until worked out in practice.
```

Comment:
This can probably be discussed as a next step, but what's the benefit of defining a CRD for this usecase? Since it's vendor specific, and the configuration will need to be consumed by the controllers, wouldn't a ConfigMap or flags on the controller not be the best way to configure default settings when installing a controller?

Reply:
For the language I noted "we are considering a pattern of using CRDs". The emphasis is on considering. This is worth digging in further on. CRDs mirrors what some others have done (e.g., cert manager) but it isn't perfect. We should hash this out and the document only notes the use case and how we are considering handling it. Something we can easily change later.

Reply:
Sounds good, thanks for the explanation

Reply:
> Something we can easily change later.

If even CRD is under question, the scope of KEP is even more unclear to me.

Reply:
@nilebox This isn't the CRD for a service. This is a separate set of CRDs to deal with controller defaults for certain conditions. There are defaults and custom settings for controllers for different providers (e.g., AWS vs Azure). You may want these cluster wide, in a namespace, or some other combination. This is for that situation rather than those related to the description of the service.